src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1528,6 +1539,11 @@ size_t ObjectSynchronizer::deflate_idle_monitors(ObjectMonitorsHashtable* table)
       // ObjectMonitors that were deflated in this cycle.
       HandshakeForDeflation hfd_hc;
       Handshake::execute(&hfd_hc);
+      // Also, we sync and desync GC threads around the handshake, so that they can
+      // safely read the mark-word and look-through to the object-monitor, without
+      // being afraid that the object-monitor is going away.
+      VM_RendezvousGCThreads sync_gc;
+      VMThread::execute(&sync_gc);
```

Comment:
Not sure why this block is under `current->is_Java_thread()` block. Shouldn't we rendezvous with GC threads regardless?

Reply:
> Not sure why this block is under `current->is_Java_thread()` block. Shouldn't we rendezvous with GC threads regardless?

Deflation is done either concurrently by the service thread, or in a safepoint by the VM thread. The current->is_Java_thread() means it's the concurrent case from the service thread. That's when it's needed. Inside of safepoints, the GC threads have already been synchronized.

Reply:
Okay, I see. Can we then move this block into separate `current->is_service_thread()`-checked block, and `assert(current->is_VM_thread())` on `else` branch? This would make this distinction very clear.